components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/model/FilterTreeBuilder.java

```diff
@@ -0,0 +1,257 @@
+                concatenatedString.append(" ").append(input.sval);
+            }
+        }
+        // Add to the list, if the filter is a simple filter.
+        if (!(concatenatedString.toString().equals(""))) {
+            tokenList.add(concatenatedString.toString());
+        }
+    }
+
+    /**
+     * Builds the binary tree from the filterString.
+     *
+     * @return the filter tree as root.
+     */
+    public Node buildTree() {
+
+        expression();
+        return root;
+    }
+
+    /**
+     * We build the parser using the recursive descent parser technique.
+     */
+    private void expression() {
+
+        term();
+        while (symbol.equals(String.valueOf("or"))) {
+            OperationNode or = new OperationNode("or");
+            or.setLeftNode(root);
+            term();
+            or.setRightNode(root);
+            root = or;
+        }
+    }
+
+    /**
+     * We build the parser using the recursive descent parser technique.
+     */
+    private void term() {
+
+        factor();
+        while (symbol.equals(String.valueOf("and"))) {
+            OperationNode and = new OperationNode("and");
+            and.setLeftNode(root);
+            factor();
+            and.setRightNode(root);
+            root = and;
+        }
+    }
+
+    /**
+     * We build the parser using the recursive descent parser technique.
+     */
+    private void factor() {
+
+        symbol = nextSymbol();
+        if (symbol.equals(String.valueOf("not"))) {
+            OperationNode not = new OperationNode("not");
+            factor();
+            not.setRightNode(root);
+            root = not;
+        } else if (symbol.equals(String.valueOf("("))) {
+            expression();
+            symbol = nextSymbol(); // We don't care about ')'.
+        } else {
+            if (!(symbol.equals(String.valueOf(")")))) {
+                ExpressionNode expressionNode = new ExpressionNode();
+                validateAndBuildFilterExpression(symbol, expressionNode);
+                root = expressionNode;
+                symbol = nextSymbol();
+            } else {
+                if (log.isDebugEnabled()) {
+                    log.debug("Invalid argument: Identity Provider Name value is empty");
+                }
+            }
+        }
+    }
+
+    /**
+     * Validate the simple filter and build a ExpressionNode
+     *
+     * @param filterString   the filter string.
+     * @param expressionNode the expression node.
+     */
+    private void validateAndBuildFilterExpression(String filterString, ExpressionNode expressionNode) {
+
+        String trimmedFilter = filterString.trim();
+        String[] filterParts;
+
+        if (Pattern.compile(Pattern.quote(" eq "), Pattern.CASE_INSENSITIVE).matcher(filterString).find()) {
+            filterParts = trimmedFilter.split(" eq | EQ | eQ | Eq ");
+            setExpressionNodeValues(filterParts[0], " eq ", filterParts[1], expressionNode);
```

Comment:
verify whether filterParts.size == 2, before accessing filterParts[1], otherwise it will result in indexoutofboundexception